src/modules/openstack.rs

```diff
@@ -0,0 +1,133 @@
+    // TODO: I'm sure there is a better way to do this.
+    let contents = match utils::read_file(&config_local) {
+        Ok(content) => {
+            //log::debug!("clouds.yaml from local directory");
+            Some(content)
+        }
+        Err(_e) => {
+            match utils::read_file(&config_user) {
+                Ok(content) => {
+                    //log::debug!("clouds.yaml from home dir");
+                    Some(content)
+                }
+                Err(_e) => {
+                    match utils::read_file(&config_global) {
+                        Ok(content) => {
+                            //log::debug!("clouds.yaml from etc");
+                            Some(content)
+                        }
+                        Err(_e) => return None,
+                    }
+                }
+            }
+        }
+    }?;
+    let clouds = YamlLoader::load_from_str(&contents).ok()?;
+    let project = &clouds[0]["clouds"][osp_cloud.unwrap()]["auth"]["project_name"]
+        .as_str()
+        .unwrap_or("");
+    if project.is_empty() {
+        return None;
+    }
+    Some(project.to_string())
+}
+
+fn get_osp_cloud_and_project(context: &Context) -> (Option<Cloud>, Option<Project>) {
+    match (
+        context.get_env("OS_CLOUD"),
+        context.get_env("OS_PROJECT_NAME"),
+    ) {
+        (Some(p), Some(r)) => (Some(p), Some(r)),
+        (None, Some(r)) => (None, Some(r)),
+        (Some(ref p), None) => (
+            Some(p.to_owned()),
+            get_osp_project_from_config(context, Some(p)),
+        ),
+        (None, None) => (None, None),
+    }
+}
+
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("openstack");
+    let config: OspConfig = OspConfig::try_load(module.config);
+
+    let (osp_cloud, osp_project) = get_osp_cloud_and_project(context);
+
+    osp_cloud.as_ref()?;
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|variable, _| match variable {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "cloud" => osp_cloud.as_ref().map(Ok),
+                "project" => osp_project.as_ref().map(Ok),
+                _ => None,
+            })
+            .parse(None)
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::error!("Error in module `openstack`: \n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::test::ModuleRenderer;
+    use ansi_term::Color;
+    use std::fs::File;
+    use std::io::{self, Write};
+
```

Comment:
Would it be possible to add a test for valid config too?

Reply:
Added.